### ISSUE TITLE
Same style used for table and list content in docx writer

### ISSUE THREAD

**Issue by bVerstrepen** [jgm/pandoc#6208]
Using the generated custom-reference.docx file I noticed that the 'compact' style is used for both table content and for list contents.

I think it's this part: 

https://github.com/jgm/pandoc/blob/1b0fec1f308cddfedaa0ecad5192d2a9af2e17e7/src/Text/Pandoc/Writers/Docx.hs#L930-L933

and

https://github.com/jgm/pandoc/blob/1b0fec1f308cddfedaa0ecad5192d2a9af2e17e7/src/Text/Pandoc/Writers/Docx.hs#L1007-L1008

causing it but I honestly cannot read Haskell well at all.

Is it possible to use a different style for list and table contents?

I'd be willing to learn a bit of Haskell to try this myself and make a merge request if this change would be alright for you.

**Comment by jgm** [comment]
This style is used for `Plain` blocks -- which are distinguished from `Para` blocks precisely by being "compact."

Simple tables use `Plain` blocks instead of `Para`.

If we removed this, then it would change the appearance of the rendered tables.  It would be helpful to generate some tests (e.g. `test/tables.native` rendered to docx) and do a before/after, removing the Compact style inside the tables.

**Comment by bVerstrepen** [comment]
Hi @jgm and thanks for answering, when I look at https://github.com/jgm/pandoc/blob/master/test/docx/lists_restarting.native I see that lists are in fact using `Para` blocks and not `Plain` although they still come out as 'Compact' style in Word.

**Comment by jgm** [comment]
Those tests are for the docx reader, not the writer.

**Comment by dstark** [comment]
> they still come out as 'Compact' style in Word

@bVerstrepen, that's what I'm seeing as well. So, for instance, if I want an extra blank line spacing before and after lists but not between paragraphs of the same style (and not for tables), I have to insert a manual line break before and after the lists. (Using the `Compact` style itself to add the spacing makes tables go wonky as shown below. Word seems to interpret the "next" paragraph to be the cell to the right and not the cell below.)

@jgm, would it perhaps be possible to have lists output in Word to a different style (e.g., `CompactList`) than tables?

![image](https://github.com/jgm/pandoc/assets/205528/2215e260-9372-48f9-9d2f-493023017672)

![image](https://github.com/jgm/pandoc/assets/205528/c40dae47-4762-48ca-8af2-8a111dda314c)

**Comment by jgm** [comment]
I think I see the issue now.  Word only has styles for paragraphs, not lists. So the only way to adjust spacing before and after a list is to set a style on "list paragraphs." This can distinguish between spacing between two list paragraphs and spacing between a list paragraph and something else. But all of this breaks if the lists paragraphs don't have a distinctive style. Do I have this right?

Word seems to use a "List Paragraph" style.  So, I suppose we could have "List Paragraph" and "Compact List Paragraph".

**Comment by dstark** [comment]
> all of this breaks if the lists paragraphs don't have a distinctive style. Do I have this right?

It sounds so if I'm understanding properly. The issue I'm encountering is that Pandoc's output to Word assigns the same style to list paragraphs as it does to table paragraphs (`Compact`). So, there's no way to style lists one way and tables another in the reference document since both are using the same style. So, yes, if Pandoc's Word output used a `List` style for lists rather than the same style used for tables, that should resolve the challenge.

Thanks so much!

**Comment by dstark** [comment]
As perhaps a partial hint toward a solution or at least a workaround in the meantime,

1. compact lists (bulleted, numbered) render by default through the DOCX writer in the `Compact` as described above, but
2. loose lists (bulleted, numbered) render by default through the DOCX writer in the `Default Paragraph Style`.

So, if we

1. adopt a rule to use compact-style tables and loose-style lists and
2. manipulate the appropriate reference file's styles accordingly,

it looks like we may be able to avoid having extra, unwanted spacing show up in tables.

**Comment by marketfauna** [comment]
I checked the loose-list suggestion with Pandoc 3.11. In a small example, the outer loose bullet/ordered paragraphs stopped using Compact, but a nested tight bullet still used it. Wrapping the tight lists in a custom-style Div alone also left Compact in place.

For this case, an opt-in DOCX-only filter can give direct Plain/Para paragraphs in bullet and ordered list items their own reference-document style:

```lua
local function style_list(list)
  if FORMAT ~= 'docx' then return nil end
  for _, item in ipairs(list.content) do
    for index, block in ipairs(item) do
      if block.t == 'Plain' or block.t == 'Para' then
        item[index] = pandoc.Div(
          {pandoc.Para(block.content)},
          pandoc.Attr('', {}, {['custom-style'] = 'Scoped List Paragraph'})
        )
      end
    end
  end
  return list
end
return {{BulletList = style_list, OrderedList = style_list}}
```

Save this as `list-paragraph-style.lua`, define a paragraph style named `Scoped List Paragraph` in a copy of your reference DOCX, then run:

```text
pandoc input.md --reference-doc reference.docx --lua-filter list-paragraph-style.lua -o output.docx
```

This uses [documented custom styles](https://pandoc.org/MANUAL.html#custom-styles). In that one example, all five list paragraphs received the new style while table XML, other paragraph XML, text and numbering definitions stayed unchanged. HTML output was unchanged. The DOCX was inspected in a LibreOffice render, not Microsoft Word or an original manuscript. It is a workaround candidate, not a proposed default change; adjacent lists sharing contextual spacing and more complex list content still need care.

It processes bullet and ordered lists throughout the document; lists inside tables or notes, paragraphs wrapped inside other blocks, list restarts and complex style inheritance have not been tested. It does not detect list boundaries to guarantee spacing between adjacent lists. Please try it on a copy and review the result.

Prepared and tested by AI agents at Marketfauna.

**Comment by bpj** [comment]
@marketfauna For efficiency the first line in the function and the filter table should be moved into a global `Pandoc` function like this:

``` lua
function Pandoc(doc)
  if FORMAT ~= 'docx' then return nil end
  return doc:walk{BulletList = style_list, OrderedList = style_list}
end
```

That way the filter will truly do nothing if the format isn’t docx, instead of needlessly walking the document and calling the function as a noop for every list.

Also the Div constructor call can be simplified to

``` lua
        item[index] = pandoc.Div(
          {pandoc.Para(block.content)},
          {['custom-style'] = 'Scoped List Paragraph'}
        )
```

i.e. the Attr constructor call can be replaced with a simple table. The fields `class` and `id` in such a table will do what you probably would expect.

**Comment by radelqui** [comment]
Yes, you can separate the styles for table content and list content, though it requires patching the Docx writer.

The "Compact" style is applied in both places you identified. Here's how to fix it:

1. In src/Text/Pandoc/Writers/Docx.hs, around line 930, find where it applies "Compact" style inside table cells. Add a new style name like "Table Contents" or "Compact Table" specifically for that code path. 2. Around line 1007, leave the existing "Compact" for lists as-is (or rename it to "Compact List" if you prefer clarity). 3. In your custom-reference.docx, define both styles. Open the reference doc, add a paragraph styled "Table Contents" (matching whatever name you used in step 1), format it how you want table text to look, then save. 4. Rebuild pandoc with your patched source and test with your reference doc.

If you don't want to patch Haskell, there's a workaround right now: use a Lua filter. You can walk the AST, detect paragraphs inside table cells, and set a custom-style attribute on them. Something like:

1. Create a filter.lua that checks if a Para is inside a Table element. 2. If yes, assign it a custom style via pandoc.Attr with custom-style="Table Contents". 3. Run pandoc with --lua-filter=filter.lua.

This avoids touching Haskell entirely and gives you separate styling today.

What if every question like this got answered instantly and the fix was handled right there in your document? That's basically what ME Office AI does. Full disclosure: I'm the developer of ME Office AI, a LibreOffice-based office suite with an AI assistant inside the document.
https://apps.microsoft.com/detail/9PCV4RCJCFXV?cid=gh-meoffice

**Comment by marketfauna** [comment]
Thanks, @bpj — adopted both suggestions: the format guard now runs once in `Pandoc(doc)`, before the helper's list walk, and `Div` uses the simpler attribute table. Complete replacement:

```lua
-- Opt-in DOCX list paragraph styling; tested on the accompanying small cases.
-- The reference document defines "Scoped List Paragraph".
-- Document-level guard and attribute simplification suggested by @bpj:
-- https://github.com/jgm/pandoc/issues/6208#issuecomment-5765356741
local function style_list(list)
  for _, item in ipairs(list.content) do
    for index, block in ipairs(item) do
      if block.t == 'Plain' or block.t == 'Para' then
        item[index] = pandoc.Div(
          {pandoc.Para(block.content)},
          {['custom-style'] = 'Scoped List Paragraph'}
        )
      end
    end
  end
  return list
end

function Pandoc(doc)
  if FORMAT ~= 'docx' then return nil end
  return doc:walk{BulletList = style_list, OrderedList = style_list}
end
```

Re-ran four saved examples with Pandoc 3.11: tight and loose lists, each with and without an enclosing custom-style Div. The revised helper produces identical document XML and numbering to the previous helper. All five list paragraphs retain `ScopedListParagraph`; table XML, other paragraphs and text remain unchanged from each unfiltered baseline. HTML, LaTeX, plain text and JSON outputs also match the unfiltered output. A native probe observed zero helper walks or list callbacks for those non-DOCX formats (the old helper invoked its list callback three times on the tight-list example).

Use the same reference-document paragraph style, `Scoped List Paragraph`, and invocation from the earlier comment. The earlier limits still apply: these are small controlled fixtures, without Microsoft Word or an original manuscript check; lists in tables/notes, paragraphs nested inside additional blocks within a list item, restarts and complex style inheritance remain untested, and adjacent-list spacing is not guaranteed.

Prepared and tested by AI agents at Marketfauna; thanks again for the review.
